src/sentry/grouping/strategies/newstyle.py

```diff
@@ -733,6 +740,83 @@ def chained_exception(
     return rv
 
 
+# See https://github.com/getsentry/rfcs/blob/main/text/0079-exception-groups.md#sentry-issue-grouping
+def filter_exceptions_for_exception_groups(
+    exceptions: List[SingleException], context: GroupingContext, event: Event, **meta: Any
+) -> List[SingleException]:
+
+    # This function only filters exceptions if there are at least two exceptions.
+    if len(exceptions) <= 1:
+        return exceptions
+
+    # Reconstruct the tree of exceptions if the required data is present.
+    nodes = {}
+    for exception in reversed(exceptions):
+        exception._children = []
+        mechanism: Mechanism = exception.mechanism
+        if mechanism and mechanism.exception_id is not None:
+            nodes[mechanism.exception_id] = exception
+            if mechanism.parent_id is not None:
+                nodes[mechanism.parent_id]._children.append(exception)
```

Comment:
If I'm following this correctly, I think it will build the list of top-level exceptions without having to reconstruct the tree - which is cool.  However I think it loses any child exceptions from each top-level exception.  Those still need to be available, so they are returned in the final result (but not used for determining distinctness).

I think I could fetch them later though.  🤔

Reply:
Also, I think this is appending exception IDs to the `top_level_exceptions` list, rather than the exceptions themselves.  But I can work through that.

Reply:
I couldn't quite make a `set` based approach work right.  We can consider this for future optimization.  Thanks.